### PR TITLE
Added the `TypedResults` option and added some tests.

### PR DESCRIPTION
**Pull request by mssola**
This is a "crazy" idea that I had not long ago. Basically, it's a bit silly that all the data fetched through the pg_fetch_* functions are strings, regardless of the type in PostgreSQL. Therefore, I've decided to add the `RespectTypes`. It's an option that defaults to false. This means that by default everything will work as always.

However, if the user sets it to `true`, then the returned types match the original types from PostgreSQL. I have included a detailed explanation in my fork [here](https://github.com/mssola/hhvm-pgsql/tree/types#postgresql-types).

Moreover, since the implementation might be a bit tricky, I've added a test for this. But, since we didn't have a proper test suite, I've created one myself :D I have explained how to run the tests also in my fork [here](https://github.com/mssola/hhvm-pgsql/tree/types#running-the-tests).

### CODE REVIEW

File: src/pgtypes.h [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+#define TRIGGEROID		2279
+#define EVTTRIGGEROID		3838
+#define LANGUAGE_HANDLEROID		2280
+#define INTERNALOID		2281
+#define OPAQUEOID		2282
+#define ANYELEMENTOID	2283
+#define ANYNONARRAYOID	2776
+#define ANYENUMOID		3500
+#define FDW_HANDLEROID	3115
+#define ANYRANGEOID		3831
```
Comment (allan-simon) [inline src/pgtypes.h]: IMHO "column" alignment should not be done with tab as you can see here in github, different settings in the size of one tab will break the alignment

I don't know what other people think ?

Comment (mssola) [inline src/pgtypes.h]: Fixed with d11660a ;-)

Comment (robbmj) [inline src/pgtypes.h]: If someone wants to volunteer their time to produce software that is
valuable I am not going to complain. Especially when the "problem" is tab
indentation. I can deal with it, functionality is what is important to me.

File: config.cmake [hunk 2]
```diff
@@ -1,12 +1,15 @@
+
 # Find include directory
+# TODO: we could just use: find_package(PostgreSQL REQUIRED). But there are
+# some Ubuntu versions with a buggy FindPostgreSQL.cmake file.
 FIND_PATH(PGSQL_INCLUDE_DIR NAMES libpq-fe.h
     PATHS /usr/include /usr/include/postgresql /usr/local/include /usr/local/include/postgresql /usr/include/pgsql)
 
-FIND_LIBRARY(PGSQL_LIBRARY NAMES pq PATHS /lib /usr/lib /usr/local/lib)
+FIND_LIBRARY(PostgreSQL_LIBRARIES NAMES pq PATHS /lib /usr/lib /usr/local/lib)
```
Comment (jrobeson) [inline config.cmake]: is there any reason to go with the mixed case here?  there's no real standard in hhvm, but there seems to be more ALL_UPPERCASE ones.

**Comment by jrobeson** [comment]
as far as the tests are concerned. we should import the ones from php and use the hhvm test runner framework. It's expected that hhvm will pull in this extension before 3.7.0 is released, so we should facilitate that happening.

**Comment by allan-simon** [comment]
"It's expected that hhvm will pull in this extension before 3.7.0 is released"

@jrobeson , sorry for my poor English, you mean  this extension will become part of the standard set of extension HHVM is packaged with? i.e it will not required to be installed separately anymore ?

**Comment by jrobeson** [comment]
@allan-simon : correct!. We were hoping this would happen for 3.6, but it doesn't seem like it will at this point.

**Comment by allan-simon** [comment]
@jrobeson : great I hope this way the blocking feature preventing my company to switch to HHVM / Hack will get more light then ( #52 ), nice to hear.

File: README.md [hunk 3]
```diff
@@ -52,6 +52,70 @@ Where `/path/to/hhvm/extensions` is a folder containing all HHVM extensions, and
...
+
+By default, all the data retrieved by pg\_fetch\_\* functions are strings. This
+is the same behavior as in the standard Zend implementation. However, you can
+change this by setting the `RespectTypes` option.
+
+~~~
+PGSQL {
+  RespectTypes = false
+}
+~~~
```
Comment (jrobeson) [inline README.md]: Unless i'm misunderstanding, the equivalent option in mysql is an ini setting called hhvm.mysql.typed_results , so you should probably rename this to TypedResults to match it.

File: config.cmake (hunk 2 above)
Comment (Norbyte) [inline config.cmake]: That's what FindPostgreSQL.cmake uses:
https://github.com/Kitware/CMake/blob/master/Modules/FindPostgreSQL.cmake

```
# ----------------------------------------------------------------------------
# The following variables are set if PostgreSQL is found:
#  PostgreSQL_FOUND         - Set to true when PostgreSQL is found.
#  PostgreSQL_INCLUDE_DIRS  - Include directories for PostgreSQL
#  PostgreSQL_LIBRARY_DIRS  - Link directories for PostgreSQL libraries
#  PostgreSQL_LIBRARIES     - The PostgreSQL libraries.
```

Comment (jrobeson) [inline config.cmake]: ah yeah of course it's gotta be this one :).  so yeah it's best to keep it like suggested then.

**Comment by jrobeson** [comment]
can you also split out the file moving from the rest of the PR?  overall this PR is not really easy to review. It's doing too many things.

**Comment by bountin** [comment]
@jrobeson @simonwelsh :+1: for bringing hhvm-pgsql into the packaged distribution - If there's anything I can help with, just get in touch!

**Comment by jrobeson** [comment]
@bountin : you could import the php pgsql and  pdo_pgsql  tests, and then filter them like hhvm does  into good, bad, slow ,etc.

**Comment by mssola** [comment]
@jrobeson First of all, thanks for reviewing it, it's been 3 months ! :)

Now, since you're saying that this extension will make it into HHVM itself, there are a couple of things to be dealt with. First of all, as you are saying, tests should be implemented as it's being done on the hhvm repo. Since doing this move can be troublesome, I'll leave it for another PR. I'd like to keep my current tests until this move has been completed though, as a reminder to implement these tests in the hhvm way.

Yes, I'll undo my / -> /src move for now. I agree that this is a matter for another PR.

File: README.md (hunk 3 above)
Comment (mssola) [inline README.md]: Ok, I didn't know about this feature. I'll rename it to be consistent with the mysql extension.

**Comment by mssola** [comment]
It's been over a year and a half, closing...

**Pull request closed without merging**